app/components/primer/octicon_component.rb

```diff
@@ -27,20 +29,94 @@ class OcticonComponent < Primer::Component
     # @param size [Symbol] <%= one_of(Primer::OcticonComponent::SIZE_MAPPINGS) %>
     # @param system_arguments [Hash] <%= link_to_system_arguments_docs %>
     def initialize(icon_name = nil, icon: nil, size: SIZE_DEFAULT, **system_arguments)
-      @icon = icon_name || icon
+      icon_key = icon_name || icon
+      cache_key = [icon_key, size, system_arguments.slice(:height, :width)].join("_")
+
       @system_arguments = system_arguments
+      @system_arguments[:tag] = :svg
+      @system_arguments[:aria] ||= {}
+
+      if @system_arguments[:aria][:label] || @system_arguments[:"aria-label"]
+        @system_arguments[:role] = "img"
+      else
+        @system_arguments[:aria][:hidden] = true
+      end
+
+      if (cache_icon = Primer::OcticonComponent::Cache.read(cache_key))
+        @icon = cache_icon
+      else
+        # Filter out classify options to prevent them from becoming invalid html attributes.
+        # Note height and width are both classify options and valid html attributes.
+        octicon_options = {
+          height: SIZE_MAPPINGS[fetch_or_fallback(SIZE_OPTIONS, size, SIZE_DEFAULT)]
+        }.merge(@system_arguments.slice(:height, :width))
 
-      @system_arguments[:class] = Primer::Classify.call(**@system_arguments)[:class]
-      @system_arguments[:height] ||= SIZE_MAPPINGS[size]
+        @icon = Octicons::Octicon.new(icon_key, octicon_options)
+        Primer::OcticonComponent::Cache.set(cache_key, @icon)
+      end
 
-      # Filter out classify options to prevent them from becoming invalid html attributes.
-      # Note height and width are both classify options and valid html attributes.
-      octicon_helper_options = @system_arguments.slice(:height, :width)
-      @system_arguments = add_test_selector(@system_arguments).except(*Primer::Classify::VALID_KEYS, :classes).merge(octicon_helper_options)
+      @system_arguments[:classes] = class_names(
+        @icon.options[:class],
+        @system_arguments[:classes]
+      )
+      @system_arguments.merge!(@icon.options.except(:class, :'aria-hidden'))
     end
 
     def call
-      octicon(@icon, { **@system_arguments })
+      render(Primer::BaseComponent.new(**@system_arguments)) { @icon.path.html_safe } # rubocop:disable Rails/OutputSafety
     end
+
+    # :nodoc:
+    class Cache
```

Comment:
could we move this to a separate file?

Reply:
Yeah I agree we should, but I'd like to have some discussion on where. I can follow up after this PR is out.